coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt

```diff
@@ -0,0 +1,276 @@
+    alignment: Alignment = Alignment.Center,
+    contentScale: ContentScale = ContentScale.Fit,
+    colorFilter: ColorFilter? = null,
+    modifier: Modifier = Modifier
+) {
+    WithConstraints(modifier) { constraints, _ ->
+        // A new GetRequest will be created on every composition, which allows the requestBuilder
+        // to work on every invocation. executeAsComposable() guards the actual execution so that
+        // the request is only run if the request changes.
+        val result = GetRequest.Builder(ContextAmbient.current)
+                .data(data)
+                .size(constraints.requestWidth.value, constraints.requestHeight.value)
+                .apply {
+                    if (requestBuilder != null) {
+                        requestBuilder(this)
+                    }
+                }
+                .build()
+                .executeAsComposable()
+
+        val resultDrawable = result?.drawable
+        if (resultDrawable != null) {
+            val painter = remember(resultDrawable) {
+                ImagePainter(resultDrawable.toBitmap().asImageAsset())
+            }
+            Image(
+                painter = painter,
+                contentScale = contentScale,
+                alignment = alignment,
+                colorFilter = colorFilter
+            )
+        }
+    }
+}
+
+/**
+ * An [ImagePainter] which draws the image with the given Android framework
+ * [android.graphics.ColorFilter].
+ */
+private class AndroidColorMatrixImagePainter(
+    private val image: ImageAsset,
+    colorFilter: android.graphics.ColorFilter
+) : Painter() {
+    private val paint = Paint()
+    private val size = PxSize(IntPx(image.width), IntPx(image.height))
+
+    init {
+        paint.asFrameworkPaint().colorFilter = colorFilter
+    }
+
+    override fun onDraw(canvas: Canvas, bounds: PxSize) {
+        // Always draw the image in the top left as we expect it to be translated and scaled
+        // in the appropriate position
+        canvas.drawImage(image, Offset.zero, paint)
+    }
+
+    /**
+     * Return the dimension of the underlying [Image] as it's intrinsic width and height
+     */
+    override val intrinsicSize: PxSize get() = size
+}
+
+/**
+ * This will execute the [GetRequest] within a composable, ensuring that the request is only
+ * execute once and storing the result, and cancelling requests as required.
+ *
+ * @return the result from the request execution, or `null` if the request has not finished yet.
+ */
+@Composable
+fun GetRequest.executeAsComposable(): RequestResult? {
+    var result by stateFor<RequestResult?>(this) { null }
+    val context = ContextAmbient.current
+
+    // Launch and execute a new request when it changes
+    onCommit(this) {
+        val job = CoroutineScope(Dispatchers.Main).launch {
+            // Start loading the image and await the result
+            result = Coil.imageLoader(context).execute(this@executeAsComposable)
```

Comment:
In the future it might make sense to offer separate singleton and non-singleton variants so users using `coil-base` can pass their own `ImageLoader` instances. Maybe `accompanist-coil-base` to follow the same naming scheme? The singleton variant could depend on the base artifact and offer `CoilImage` overloads that pass the singleton `ImageLoader` by default.

Reply:
SGTM, but I'll leave this as a task for when we upstream back to the Coil library.